### PR TITLE
Converted classic pipeline to yaml - AWS Publish Images 3.0

### PR DESCRIPTION
**Pull request by priyadishah**
Converted classic pipeline to YAML https://github.com/robe070/cookbooks/issues/184

Classic Pipeline
Repository: robe070/cookbooks
Branch: debug/paas
Release Pipeline Name - Image Release 2.0

Converted classic to yaml pipeline
Repository: priyadishah/cookbooks
Branch: feature/AWSImageRelease-3.0
Pipeline Name: AWS Publish Images 3.0
Pipeline Url: https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build?definitionId=68

### CODE REVIEW

**Comment by priyadishah** [comment]
@robe070 while listing AMIs, that part is showing as Skip in "Update Template AMI Mappings" step., am I missing anything? 

https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2278&view=logs&j=6cf04ae0-b4bb-5ee8-b781-bd6226e1f450&t=98bd6c20-9950-5129-8a29-9eae8b1ee588&l=67

**Comment by priyadishah** [comment]
@robe070 I have separated templates file based on stages, how can I put to AzureDevOps folder and how can I differentiate it? Please advise.

-Mohit

**Review by robe070 — CHANGES_REQUESTED** [review]
I do not know what TagGit.yml is doing here. Tagging will be done in the cloudformation templates in the lansa/aws-templates repo. And this is about Images. You probably need to complete the Publish Templates 3.0 pipeline if the Publish Images pipeline uses the cloudformation templates.

**Review by robe070 — COMMENTED** [review]
Your Templates folder path is wrong. Top level is AzureDevOps\templates.

And I've changed the names. Talk to @nkhl-accolite

**Comment by robe070** [comment]
> @robe070 I have separated templates file based on stages, how can I put to AzureDevOps folder and how can I differentiate it? Please advise.
> 
> -Mohit

How? create the folder and move the templates folder into it!

**Comment by robe070** [comment]
> @robe070 while listing AMIs, that part is showing as Skip in "Update Template AMI Mappings" step., am I missing anything?
> 
> https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2278&view=logs&j=6cf04ae0-b4bb-5ee8-b781-bd6226e1f450&t=98bd6c20-9950-5129-8a29-9eae8b1ee588&l=67

It should not say `skip`.
It means it has not located a matching image name.
Look in the script - `UpdateTemplateAMIMappings.ps1`. It might be easiest to improve the logging in that script. Change line 64 and onwards from this:

```
        if ( Test-Path $path ) {
            $amiID = [string](Get-Content -Path $path)
            $AMIList += $amiID
        } else {
            $AMIList += "skip"
        }
```
to this:
```
        Write-Host("Path to be tested: $path")
        if ( Test-Path $path ) {
            $amiID = [string](Get-Content -Path $path)
            $AMIList += $amiID
        } else {
            Write-Host("Path does not exist.")
            $AMIList += "skip"
        }
```
Is this another 'S' issue?
cc: @nkhl-accolite

**Comment by priyadishah** [comment]
> > @robe070 while listing AMIs, that part is showing as Skip in "Update Template AMI Mappings" step., am I missing anything?
> > https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2278&view=logs&j=6cf04ae0-b4bb-5ee8-b781-bd6226e1f450&t=98bd6c20-9950-5129-8a29-9eae8b1ee588&l=67
> 
> It should not say `skip`. It means it has not located a matching image name. Look in the script - `UpdateTemplateAMIMappings.ps1`. It might be easiest to improve the logging in that script. Change line 64 and onwards from this:
> 
> ```
>         if ( Test-Path $path ) {
>             $amiID = [string](Get-Content -Path $path)
>             $AMIList += $amiID
>         } else {
>             $AMIList += "skip"
>         }
> ```
> 
> to this:
> 
> ```
>         Write-Host("Path to be tested: $path")
>         if ( Test-Path $path ) {
>             $amiID = [string](Get-Content -Path $path)
>             $AMIList += $amiID
>         } else {
>             Write-Host("Path does not exist.")
>             $AMIList += "skip"
>         }
> ```
> 
> Is this another 'S' issue? cc: @nkhl-accolite

Correct but without 'S', it's unable to checkout Git Repo that's why I added 'S'

**Comment by robe070** [comment]
AzureDevOps issue. Can you specify the path for the artefact?

**Comment by robe070** [comment]
Clearly the YAML conversion has changed the default download directory. this link explains the directory structure:
https://dev.to/shleeen/understanding-the-directory-structure-created-by-azure-devops-tasks-1328

**Review by robe070 — CHANGES_REQUESTED** [review]
Please fix your template path and template names right now!
cc: @nkhl-accolite

**Comment by priyadishah** [comment]
> > @robe070 while listing AMIs, that part is showing as Skip in "Update Template AMI Mappings" step., am I missing anything?
> > https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2278&view=logs&j=6cf04ae0-b4bb-5ee8-b781-bd6226e1f450&t=98bd6c20-9950-5129-8a29-9eae8b1ee588&l=67
> 
> It should not say `skip`. It means it has not located a matching image name. Look in the script - `UpdateTemplateAMIMappings.ps1`. It might be easiest to improve the logging in that script. Change line 64 and onwards from this:
> 
> ```
>         if ( Test-Path $path ) {
>             $amiID = [string](Get-Content -Path $path)
>             $AMIList += $amiID
>         } else {
>             $AMIList += "skip"
>         }
> ```
> 
> to this:
> 
> ```
>         Write-Host("Path to be tested: $path")
>         if ( Test-Path $path ) {
>             $amiID = [string](Get-Content -Path $path)
>             $AMIList += $amiID
>         } else {
>             Write-Host("Path does not exist.")
>             $AMIList += "skip"
>         }
> ```
> 
> Is this another 'S' issue? cc: @nkhl-accolite

the above code worked and I can see listing of AMIs,

**Comment by priyadishah** [comment]
> AzureDevOps issue. Can you specify the path for the artefact?

@robe070 we have removed "S/" from path and it is resolved. All artefects and AMIs list are being downloaded under same path.
image

**Comment by priyadishah** [comment]
@robe070 in "w16d-14-2 IMG Test" job at Deploy AWS template step, I am getting error of "##[error]ValidationError: Parameter '11WebserverOSVersion' must be one of AllowedValues" https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2398&view=logs&j=09659086-167b-54fe-bc28-78372387c7b5&t=b402e615-8832-558c-2457-acbbe8a7eadb&l=206

**Comment by priyadishah** [comment]
Also @robe070 do we need to add AMI listing of Windows 2022 & Windows 2022 JPN under "Update Template AMI Mappings" jobs?
https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2398&view=logs&j=6cf04ae0-b4bb-5ee8-b781-bd6226e1f450&t=98bd6c20-9950-5129-8a29-9eae8b1ee588&l=75

**Comment by robe070** [comment]
> @robe070 in "w16d-14-2 IMG Test" job at Deploy AWS template step, I am getting error of "##[error]ValidationError: Parameter '11WebserverOSVersion' must be one of AllowedValues" https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2398&view=logs&j=09659086-167b-54fe-bc28-78372387c7b5&t=b402e615-8832-558c-2457-acbbe8a7eadb&l=206

It looks like the parameter name should be 'WebserverOSVersion'. Your team can help you if you do not understand CloudFormation templates.

**Comment by robe070** [comment]
> Also @robe070 do we need to add AMI listing of Windows 2022 & Windows 2022 JPN under "Update Template AMI Mappings" jobs? https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2398&view=logs&j=6cf04ae0-b4bb-5ee8-b781-bd6226e1f450&t=98bd6c20-9950-5129-8a29-9eae8b1ee588&l=75

Yes you do and you need to add those AMIs into the cloudformation template

**Comment by priyadishah** [comment]
> > Also @robe070 do we need to add AMI listing of Windows 2022 & Windows 2022 JPN under "Update Template AMI Mappings" jobs? https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2398&view=logs&j=6cf04ae0-b4bb-5ee8-b781-bd6226e1f450&t=98bd6c20-9950-5129-8a29-9eae8b1ee588&l=75
> 
> Yes you do and you need to add those AMIs into the cloudformation template

alright, I am checking and I will add values of windows 2022.

**Comment by priyadishah** [comment]
> > @robe070 in "w16d-14-2 IMG Test" job at Deploy AWS template step, I am getting error of "##[error]ValidationError: Parameter '11WebserverOSVersion' must be one of AllowedValues" https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2398&view=logs&j=09659086-167b-54fe-bc28-78372387c7b5&t=b402e615-8832-558c-2457-acbbe8a7eadb&l=206
> 
> It looks like the parameter name should be 'WebserverOSVersion'. Your team can help you if you do not understand CloudFormation templates.

sure, I am checking with Nikhil.

**Comment by priyadishah** [comment]
@robe070 can you please guide me here? Stack creation request failed with error: 'Stack w16d-15-0-IMG-w16d-15-0 failed to reach completion status, error: 'Resource is not in the state stackCreateComplete'' Error: Stack w16d-15-0-IMG-w16d-15-0 failed to reach completion status, error: 'Resource is not in the state stackCreateComplete'

https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2408&view=logs&j=00fc5acc-a0ff-5dc7-cc25-1d4880bc49eb&t=da8abc74-a97f-5fd9-5fc5-54fad5c2aa1b&l=52

**Comment by robe070** [comment]
> @robe070 can you please guide me here? Stack creation request failed with error: 'Stack w16d-15-0-IMG-w16d-15-0 failed to reach completion status, error: 'Resource is not in the state stackCreateComplete'' Error: Stack w16d-15-0-IMG-w16d-15-0 failed to reach completion status, error: 'Resource is not in the state stackCreateComplete'
> 
> https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2408&view=logs&j=00fc5acc-a0ff-5dc7-cc25-1d4880bc49eb&t=da8abc74-a97f-5fd9-5fc5-54fad5c2aa1b&l=52

I looked at w16d-14-2-IMG
Its parameters are:
```
win2019 => This is wrong.
Marketplace => This needs to be checked. I suspect its wrong
V14 SP2
```
Same for w16d-15-0-IMG
```
win2019 => This is wrong.
Marketplace => This needs to be checked. I suspect its wrong
V14 SP2 => This is wrong
```

I conclude that you are not passing parameters to the DevOps yaml template because these values do not change and they are just about the only thing that needs to change between each stage.

Check all these parameters in the Classic pipeline. Ensure they are correct in the YAML pipeline. If and only if you have thoroughly compared them and there are still errors, come back to me.
@nkhl-accolite

**Comment by priyadishah** [comment]
> > @robe070 can you please guide me here? Stack creation request failed with error: 'Stack w16d-15-0-IMG-w16d-15-0 failed to reach completion status, error: 'Resource is not in the state stackCreateComplete'' Error: Stack w16d-15-0-IMG-w16d-15-0 failed to reach completion status, error: 'Resource is not in the state stackCreateComplete'
> > https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2408&view=logs&j=00fc5acc-a0ff-5dc7-cc25-1d4880bc49eb&t=da8abc74-a97f-5fd9-5fc5-54fad5c2aa1b&l=52
> 
> I looked at w16d-14-2-IMG Its parameters are:
> 
> ```
> win2019 => This is wrong.
> Marketplace => This needs to be checked. I suspect its wrong
> V14 SP2
> ```
> 
> Same for w16d-15-0-IMG
> 
> ```
> win2019 => This is wrong.
> Marketplace => This needs to be checked. I suspect its wrong
> V14 SP2 => This is wrong
> ```
> 
> I conclude that you are not passing parameters to the DevOps yaml template because these values do not change and they are just about the only thing that needs to change between each stage.
> 
> Check all these parameters in the Classic pipeline. Ensure they are correct in the YAML pipeline. If and only if you have thoroughly compared them and there are still errors, come back to me. @nkhl-accolite

Hi @robe070 

@nkhl-accolite and me have verified and updated parameters in DevOps yaml file and also i have compared with Classic Pipeline as well. Apart from this, I have added values of "Win 2022" in UpdateTemplateAMIMappings.ps1 file & lansa master templates.

Also, we have added AWS Tagging in lansa-master-win.cfn.template, webserver-win.cfn.template & lansa-stack-type-win.cfn.template and getting below error.

https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2513&view=logs&j=2659fff2-e774-5509-aff6-22c42f71e24d&t=07a032f3-f620-5b0d-6750-0844f496cf5d&l=52

File: scripts/azure_set_gate_variable_cloud_account_id.ps1 [hunk 1]
```diff
@@ -45,5 +45,6 @@ if (Test-Path $path) {
     Write-Host "##vso[task.setvariable variable=Sku;isOutput=true]$sku"
     Write-host "The value of Variable IsEnabled is updated to True and output variable ImageUrl to $uri"
 } else {
-    Write-Host "Artifact path $path does NOT exist for $Version"
+    #Write-Host "Artifact path $path does NOT exist for $Version"
+    throw "Artifact path does NOT exist for $Version" # Throwing error if there's no baseimage, instead of just writing it to the host.
```
Comment (robe070) [inline scripts/azure_set_gate_variable_cloud_account_id.ps1]: This is good coding. But remove all the comments - line 48 entirely and the comment on the end of line 49. It doesn't add anything. Its standard POSH coding at LANSA.

The same for the other script.

**Review by robe070 — CHANGES_REQUESTED** [review]

**Comment by robe070** [comment]
> Also, we have added AWS Tagging in lansa-master-win.cfn.template, webserver-win.cfn.template & lansa-stack-type-win.cfn.template and getting below error.
> 
> https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2513&view=logs&j=2659fff2-e774-5509-aff6-22c42f71e24d&t=07a032f3-f620-5b0d-6750-0844f496cf5d&l=52

I commented on the above in Teams and in the weekly meeting.

I also mentioned that until the tagging of the resources has been separately completed, I will not be reviewing the details of this pipeline as the tagging may well change the results.

**Comment by priyadishah** [comment]
> > @robe070 in "w16d-14-2 IMG Test" job at Deploy AWS template step, I am getting error of "##[error]ValidationError: Parameter '11WebserverOSVersion' must be one of AllowedValues" https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2398&view=logs&j=09659086-167b-54fe-bc28-78372387c7b5&t=b402e615-8832-558c-2457-acbbe8a7eadb&l=206
> 
> It looks like the parameter name should be 'WebserverOSVersion'. Your team can help you if you do not understand CloudFormation templates.

this issue has been resolved on the same day. Adding comments now.

**Comment by priyadishah** [comment]
> > Also, we have added AWS Tagging in lansa-master-win.cfn.template, webserver-win.cfn.template & lansa-stack-type-win.cfn.template and getting below error.
> > https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2513&view=logs&j=2659fff2-e774-5509-aff6-22c42f71e24d&t=07a032f3-f620-5b0d-6750-0844f496cf5d&l=52
> 
> I commented on the above in Teams and in the weekly meeting.
> 
> I also mentioned that until the tagging of the resources has been separately completed, I will not be reviewing the details of this pipeline as the tagging may well change the results.

Hi @robe070  we can see tagging on AWS resources on the same day.
image
image
image

**Comment by priyadishah** [comment]
Hi @robe070 There is a step for Git in which it pushes/merges our feature aws-template branch (https://github.com/priyadishah/aws-templates-md/tree/support/scalable) to original aws-template branch (https://github.com/lansa/aws-templates) and at that step, I am getting an error: 

https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2690&view=logs&j=71d46a05-0502-5531-8fd1-b0dff1ab9f27&t=fd191462-7c69-53a9-3a6d-c359ecd6d640&l=69

@nkhl-accolite @priyadishah

**Comment by robe070** [comment]
> Hi @robe070 we can see tagging on AWS resources on the same day.

What is this meant to mean? Demonstrate the tags are applied? I agree that has been done.

**Comment by robe070** [comment]
> Hi @robe070 There is a step for Git in which it pushes/merges our feature aws-template branch (https://github.com/priyadishah/aws-templates-md/tree/support/scalable) to original aws-template branch (https://github.com/lansa/aws-templates) and at that step, I am getting an error:
> 
> https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2690&view=logs&j=71d46a05-0502-5531-8fd1-b0dff1ab9f27&t=fd191462-7c69-53a9-3a6d-c359ecd6d640&l=69
> 
@nkhl-accolite

Solution is on the next line!
https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2690&view=logs&j=71d46a05-0502-5531-8fd1-b0dff1ab9f27&t=fd191462-7c69-53a9-3a6d-c359ecd6d640&l=70

**Comment by priyadishah** [comment]
> > Hi @robe070 There is a step for Git in which it pushes/merges our feature aws-template branch (https://github.com/priyadishah/aws-templates-md/tree/support/scalable) to original aws-template branch (https://github.com/lansa/aws-templates) and at that step, I am getting an error:
> > https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2690&view=logs&j=71d46a05-0502-5531-8fd1-b0dff1ab9f27&t=fd191462-7c69-53a9-3a6d-c359ecd6d640&l=69
> 
> @nkhl-accolite
> 
> Solution is on the next line! https://dev.azure.com/VisualLansa/Lansa%20AWS%20Scalable%20License%20Images/_build/results?buildId=2690&view=logs&j=71d46a05-0502-5531-8fd1-b0dff1ab9f27&t=fd191462-7c69-53a9-3a6d-c359ecd6d640&l=70

Right @robe070 but I have already open PR from this branch so I can't take any action on it.
@nkhl-accolite

**Comment by robe070** [comment]
> Right @robe070 but I have already open PR from this branch so I can't take any action on it.

Rebase on the lansa repo

File: scripts/azure_set_gate_variable_cloud_account_id.ps1 [hunk 2]
```diff
@@ -45,6 +45,5 @@ if (Test-Path $path) {
     Write-Host "##vso[task.setvariable variable=Sku;isOutput=true]$sku"
     Write-host "The value of Variable IsEnabled is updated to True and output variable ImageUrl to $uri"
 } else {
-    #Write-Host "Artifact path $path does NOT exist for $Version"
-    throw "Artifact path does NOT exist for $Version" # Throwing error if there's no baseimage, instead of just writing it to the host.
+    Write-Host "Artifact path $path does NOT exist for $Version"
```
Comment (robe070) [inline scripts/azure_set_gate_variable_cloud_account_id.ps1]: Why have you removed the throw?

**Review by robe070 — CHANGES_REQUESTED** [review]

File: scripts/azure_set_gate_variable_cloud_account_id.ps1 (hunk 2 above)
Comment (priyadishah) [reply scripts/azure_set_gate_variable_cloud_account_id.ps1]: Hi @robe070 where do you see this code? I added this but removed it very next day, I verified twice and couldn't see on both feature branch. this is must be old.

**Comment by robe070** [comment]
Please resolve the conflicts ASAP so I can merge this PR.

**Comment by priyadishah** [comment]
> Please resolve the conflicts ASAP so I can merge this PR.

Hi @robe070 I have resolved conflict.

**Comment by priyadishah** [comment]
> > Please resolve the conflicts ASAP so I can merge this PR.
> 
> Hi @robe070 I have resolved conflict.

Hi @robe070 if this is all good, can you please merge this?

**Pull request merged**